src/core/pointcloud/qgspointcloudlayerexporter.cpp

```diff
@@ -0,0 +1,585 @@
+  }
+}
+
+void QgsPointCloudLayerExporter::doExport()
+{
+  mTransform = new QgsCoordinateTransform( mSourceCrs, mTargetCrs, mTransformContext );
+  if ( mExtent.isFinite() )
+  {
+    try
+    {
+      mExtent = mTransform->transformBoundingBox( mExtent, Qgis::TransformDirection::Reverse );
+    }
+    catch ( const QgsCsException &cse )
+    {
+      QgsDebugMsg( QStringLiteral( "Error transforming extent: %1" ).arg( cse.what() ) );
+    }
+  }
+
+  if ( mFormat == QLatin1String( "memory" ) )
+  {
+    if ( !mMemoryLayer )
+      prepareExport();
+
+    ExporterMemory exp = ExporterMemory( this );
+    exp.run();
+  }
+#ifdef HAVE_PDAL_QGIS
+  else if ( mFormat == QLatin1String( "LAZ" ) )
+  {
+    setAllAttributes();
+    // PDAL may throw exceptions
+    try
+    {
+      ExporterPdal exp = ExporterPdal( this );
+      exp.run();
+    }
+    catch ( std::runtime_error &e )
+    {
+      setLastError( QString::fromLatin1( e.what() ) );
+      QgsDebugMsg( QStringLiteral( "PDAL has thrown an exception: {}" ).arg( e.what() ) );
+    }
+  }
+#endif
+  else
+  {
+    QgsVectorFileWriter::SaveVectorOptions saveOptions;
+    saveOptions.layerName = mName;
+    saveOptions.driverName = mFormat;
+    saveOptions.datasourceOptions = QgsVectorFileWriter::defaultDatasetOptions( mFormat );
+    saveOptions.layerOptions = QgsVectorFileWriter::defaultLayerOptions( mFormat );
+    saveOptions.symbologyExport = QgsVectorFileWriter::NoSymbology;
+    saveOptions.actionOnExistingFile = mActionOnExistingFile;
+    saveOptions.feedback = mFeedback;
+    mVectorSink = QgsVectorFileWriter::create( mFilename, outputFields(), QgsWkbTypes::PointZ, mTargetCrs, QgsCoordinateTransformContext(), saveOptions );
+    ExporterVector exp = ExporterVector( this );
+    exp.run();
+  }
+}
+
+QgsMapLayer *QgsPointCloudLayerExporter::getExportedLayer()
+{
+  if ( mFormat == QLatin1String( "memory" ) && mMemoryLayer )
+  {
+    QgsMapLayer *retVal = mMemoryLayer;
+    mMemoryLayer = nullptr;
+    return retVal;
+  }
+
+  const QFileInfo fileInfo( mFilename );
+
+  if ( mFormat == QLatin1String( "LAZ" ) )
+  {
+    return new QgsPointCloudLayer( mFilename, fileInfo.completeBaseName(), QStringLiteral( "pdal" ) );
+  }
+
+  if ( mFormat == QLatin1String( "GPKG" ) )
+  {
+    QString uri( mFilename );
+    uri += "|layername=" + mName;
+    return new QgsVectorLayer( uri, mName, QStringLiteral( "ogr" ) );
+  }
+
+  return new QgsVectorLayer( mFilename, fileInfo.completeBaseName(), QStringLiteral( "ogr" ) );
+}
+
+//
+// ExporterBase
+//
+
+void QgsPointCloudLayerExporter::ExporterBase::run()
+{
+  QVector<IndexedPointCloudNode> nodes;
+  qint64 pointCount = 0;
+  QQueue<IndexedPointCloudNode> queue;
+  queue.push_back( mParent->mIndex->root() );
+  while ( !queue.empty() )
+  {
+    IndexedPointCloudNode node = queue.front();
+    queue.pop_front();
+    if ( !nodes.contains( node ) &&
```

Comment:
why do we need to check whether `nodes` contains `node`? each not should only be visited once, or not?

Reply:
Hmm, you're right. This was copied from QgsPointCloudStatsCalculator, but the logic is simpler and this check is unneeded.